slave-new/process-voms/lib/process-voms.pl

```diff
@@ -0,0 +1,255 @@
+			@{$rolesToDismiss{"$group"}{"$role"}} = array_minus_deep(@{$rolesToDismiss{"$group"}{"$role"}}, @{$membersToRemove{"/$name"}}); # No need to revoke roles if the user is going to be fully removed
+			@{$rolesToDismiss{"$group"}{"$role"}} = array_minus_deep(@{$rolesToDismiss{"$group"}{"$role"}}, @{$membersToRemove{"$group"}}); # No need to revoke roles if the user is going to be removed from the group
+		}
+        }
+
+
+	# Effect changes
+	# 1. create / delete groups
+	foreach $group (@groupsToDelete) {
+		effectCall "voms-admin --vo $name delete-group \"$group\"",
+		"deleting Group \"$group\" from VO \"$name\"";
+	}
+	foreach $group (@groupsToCreate) {
+		effectCall "voms-admin --vo $name create-group \"$group\"",
+		"creating Group \"$group\" in VO \"$name\"";
+	}
+
+	# 2. create / delete roles
+	foreach $role (@rolesToDelete) {
+		effectCall "voms-admin --vo $name delete-role \"$role\"",
+		"deleting Role \"$role\" from VO \"$name\".";
+	}
+	foreach $role (@rolesToCreate) {
+		effectCall "voms-admin --vo $name create-role \"$role\"",
+		"creating Role \"$role\" in VO \"$name\".";
+	}
+
+	# 3. add members to/remove members from groups
+	foreach $group (@groups_toBe) {
+		foreach $user (@{$membersToRemove{"$group"}}) {
+			if( "$group" eq "/$name" ) { # Root group?
+				effectCall "voms-admin --nousercert --vo $name delete-user \"$user->{'DN'}\" \"$user->{'CA'}\"",
+				"deleting user \"$user->{'DN'}\" from VO \"$name\".";
+			}
+			else {
+				effectCall "voms-admin --nousercert --vo $name remove-member \"$group\" \"$user->{'DN'}\" \"$user->{'CA'}\"",
+				"removing user \"$user->{'DN'}\" from Group \"$group\" in VO \"$name\".";
+			}
+		}
+		foreach $user (@{$membersToAdd{"$group"}}) {
+			if( "$group" eq "/$name" ) { # Root group?
+				effectCall "voms-admin --nousercert --vo $name create-user \"$user->{'DN'}\" \"$user->{'CA'}\" \"$user->{'CN'}\" \"$user->{'email'}\"",
+				"creating user \"$user->{'DN'}\" in VO \"$name\".";
+			}
+			else {
+				effectCall "voms-admin --nousercert --vo $name add-member \"$group\" \"$user->{'DN'}\" \"$user->{'CA'}\"",
+				"adding user \"$user->{'DN'}\" to Group \"$group\" in VO \"$name\".";
+			}
+		}
+	}
+
+	# 4. assign/dismiss roles
+	foreach $group (@groups_toBe) {
+		foreach $role (@roles_toBe) {
+			foreach $user (@{$rolesToDismiss{"$group"}{"$role"}}) {
+				effectCall "voms-admin --nousercert --vo $name dismiss-role \"$group\" \"$role\" \"$user->{'DN'}\" \"$user->{'CA'}\"",
+				"stripping user \"$user->{'DN'}\" of Role \"$role\" for Group \"$group\" in VO \"$name\"";
+			}
+			foreach $user (@{$rolesToAssign{"$group"}{"$role"}}) {
+				effectCall "voms-admin --nousercert --vo $name assign-role \"$group\" \"$role\" \"$user->{'DN'}\" \"$user->{'CA'}\"",
+				"assigning Role \"$role\" to user \"$user->{'DN'}\" for Group \"$group\" in VO \"$name\"";
+			}
+		}
+	}
+}
+
+closelog();
+
+$retval;
```

Comment:
Is this correct? Variable retval is created on the line 98 but then nothing set it with any value, just return at the end of the script.